### PR TITLE
feat: wasm32-wasi support

### PR DESCRIPTION
**Pull request by rjzak**
Uses RustCrypto, tests pass with both Wasmtime and Enarx!

Likely there will be requested changes to `.cargo/config` around the default target and preferred WebAssembly runtime.

Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -225,6 +247,44 @@ async fn main() -> anyhow::Result<()> {
     Ok(())
 }
 
+#[cfg(target_arch = "wasm32")]
```
Comment (rvolosatovs) [inline src/main.rs:250]: We should not duplicate code like this, instead, introduce a function for specific things that are different and switch on the architecture.

File: src/main.rs [hunk 2]
```diff
@@ -191,10 +191,32 @@ impl State {
         let crt = tbs.sign(&pki)?;
         Ok(Self { key, crt, san })
     }
+
+    #[cfg(target_arch = "wasm32")]
```
Comment (rvolosatovs) [inline src/main.rs:195]: This is only used for testing and does not belong in production code

File: src/main.rs (hunk 1 above)
Comment (rjzak) [reply src/main.rs:250]: Two options:
* Use the code I wrote, which changes the `tokio` runtime based on Wasi or native.
* Force native versions of Steward to use single-threaded `tokio`.

File: .cargo/config [hunk 3]
```diff
@@ -0,0 +1,4 @@
+[target.wasm32-wasi]
+rustflags = ["--cfg", "tokio_unstable"]
+#runner = ["enarx", "run", "--wasmcfgfile", "Enarx.toml"]
```
Comment (rvolosatovs) [inline .cargo/config:3]: Aren't we supposed to run this in Enarx? Why is this necessary, does this not work in Enarx? If so, what's the purpose of this PR?

File: src/main.rs (hunk 2 above)
Comment (rjzak) [reply src/main.rs:195]: Where should it go then? Maybe a decorator to only compile it for testing? Also, this addresses today's lack of file system support in Enarx.

File: .cargo/config (hunk 3 above)
Comment (rjzak) [reply .cargo/config:3]: It works in Enarx, but Wasmtime is:
* Faster, and
* Presumably easier to get loaded in a CI environment.

File: src/main.rs (hunk 1 above)
Comment (rvolosatovs) [reply src/main.rs:245]: Discussed offline (it should be possible to use `cfg_attr`)

File: Cargo.toml [hunk 4]
```diff
@@ -1,3 +1,9 @@
+[patch.crates-io]
+#tokio = { git = "https://github.com/rjzak/tokio", branch = "master" }
+#tokio-macros = { git = "https://github.com/rjzak/tokio", branch = "master" }
+#tokio-util = { git = "https://github.com/rjzak/tokio", branch = "master" }
```
Comment (rvolosatovs) [inline Cargo.toml:4]: There should either be a comment about why this is commented-out with a reference to a technical debt issue or these commented-out blocks should not be here

File: Cargo.toml [hunk 5]
```diff
@@ -28,8 +34,9 @@ tracing-subscriber = { version="^0.3.15", features = ["env-filter", "json"] }
 tower-http = { version = "^0.3.0", features = ["trace"] }
 axum = { version = "^0.5.16", features = ["headers"] }
 clap = { version = "^3.0.14", features = ["derive", "env"] }
-hyper = { version = "^0.14.17", features = ["full"] }
-tokio = { version = "^1.15.0", features = ["full"] }
+# hyper = { version = "^0.14.20", features = ["full"] }
```
Comment (rvolosatovs) [inline Cargo.toml:37]: See above

File: Cargo.toml [hunk 6]
```diff
@@ -28,8 +34,9 @@ tracing-subscriber = { version="^0.3.15", features = ["env-filter", "json"] }
...
+hyper = { git = "https://github.com/rjzak/hyper", branch = "wasi_wip", features = ["server"] }
+tokio = { version = "^1.21.2", features = ["full"] }
```
Comment (rvolosatovs) [inline Cargo.toml:39]: Do we really need `"full"` or is only a subset of features enough to cover our use case?

File: .cargo/config (hunk 3 above)
Comment (rjzak) [reply .cargo/config]: As per our conversation, I've removed the reference to Wasmtime, and changed the CI config to download and install Enarx, and confirmed that this works.

File: src/main.rs (hunk 2 above)
Comment (rjzak) [reply src/main.rs]: Resolved, no test data is ever inadvertently added to production code.

File: src/main.rs (hunk 1 above)
Comment (rjzak) [reply src/main.rs]: Resolved, `cfg_attr` is used.

File: src/main.rs [hunk 7]
```diff
@@ -193,7 +207,8 @@ impl State {
     }
 }
 
-#[tokio::main]
+#[cfg_attr(target_os = "wasi", tokio::main(flavor = "current_thread"))]
+#[cfg_attr(any(target_family = "unix", windows), tokio::main)]
```
Comment (rvolosatovs) [inline src/main.rs:211]:
```suggestion
#[cfg_attr(not(target_os = "wasi"), tokio::main)]
```

File: src/main.rs [hunk 8]
```diff
@@ -135,6 +135,20 @@ impl State {
...
+        PrivateKeyInfo::from_der(key)?;
+        Certificate::from_der(crt)?;
+
+        Ok(State {
+            crt: crt.to_vec(),
+            san: None,
+            key: key.to_vec().into(),
+        })
+    }
+
```
Comment (rvolosatovs) [inline src/main.rs:151]: I think what we want is `read(san: Option<String>, mut key: impl BufRead, mut crt: impl BufRead)`

That method would not need to be guarded by either `target_os` or `test`, instead - `State::load` should use that method directly after opening the files and constructing the buffered readers
This way we minimize the amount of duplication, amount of code we need to maintain *and* increase test coverage. (Note that parsing `pem` in the test would also be unnecessary)

**Review by rvolosatovs — COMMENTED** [review]
This looks much better than before and is definitely moving the right direction!

File: Cargo.toml (hunk 6 above)
Comment (rvolosatovs) [reply Cargo.toml:35]: Could you file an issue to revise the list of features we enable in our dependencies and to remove redundant dependencies?

File: deny.toml [hunk 9]
```diff
@@ -16,3 +16,6 @@ allow = [
     "BSD-3-Clause",
     "Unicode-DFS-2016",
 ]
+
+[sources.allow-org]
+github = ["rjzak"] # temporary until Hyper PR #2900 is merged
```
Comment (rvolosatovs) [inline deny.toml:21]: This should be URL to the PR

File: src/main.rs [hunk 10]
```diff
@@ -217,10 +228,26 @@ async fn main() -> anyhow::Result<()> {
...
+        axum::Server::bind(&addr)
+            .serve(app(state).into_make_service())
+            .await?;
+    }
+    #[cfg(target_os = "wasi")]
+    {
+        use std::os::wasi::io::FromRawFd;
+        tracing::debug!("listening");
+        let std_listener = unsafe { std::net::TcpListener::from_raw_fd(3) };
+        std_listener.set_nonblocking(true).unwrap();
```
Comment (rvolosatovs) [inline src/main.rs:245]:
```suggestion
        std_listener.set_nonblocking(true).context("failed to set NONBLOCK")?;
```

File: src/main.rs [hunk 11]
```diff
@@ -217,10 +228,26 @@ async fn main() -> anyhow::Result<()> {
...
+        axum::Server::from_tcp(std_listener)
+            .unwrap()
```
Comment (rvolosatovs) [inline src/main.rs:247]:
```suggestion
            .context("failed to construct server")?
```

File: src/main.rs (hunk 8 above)
Comment (rjzak) [reply src/main.rs]: Done

File: src/main.rs [hunk 12]
```diff
@@ -460,7 +487,15 @@ mod tests {
         use tower::ServiceExt; // for `app.oneshot()`
 
         fn certificates_state() -> State {
-            State::load(None, "testdata/ca.key", "testdata/ca.crt").unwrap()
+            #[cfg(not(target_os = "wasi"))]
+            return State::load(None, "testdata/ca.key", "testdata/ca.crt").unwrap();
```
Comment (rvolosatovs) [inline src/main.rs:491]:
```suggestion
            return State::load(None, "testdata/ca.key", "testdata/ca.crt").expect("failed to load state");
```

File: src/main.rs [hunk 13]
```diff
@@ -460,7 +487,15 @@ mod tests {
...
+            #[cfg(target_os = "wasi")]
+            {
+                let crt = std::io::BufReader::new(include_bytes!("../testdata/ca.crt").as_slice());
+                let key = std::io::BufReader::new(include_bytes!("../testdata/ca.key").as_slice());
+
+                State::read(None, key, crt).unwrap()
```
Comment (rvolosatovs) [inline src/main.rs:497]:
```suggestion
                State::read(None, key, crt).expect("failed to read state")
```

**Review by rvolosatovs — COMMENTED** [review]
Nice progress!

File: src/main.rs [hunk 14]
```diff
@@ -10,8 +10,9 @@ mod crypto;
 mod ext;
 
 use crypto::*;
+use std::io::BufRead;
```
Comment (rvolosatovs) [inline src/main.rs:13]: this is in the wrong import group

File: Cargo.toml (hunk 6 above)
Comment (rjzak) [reply Cargo.toml]: Full was there because that was in Cargo.toml when I began working on Wasi support. I've since reduced it to `rt` and `macros`, and `rt-multi-threaded` for non-Wasi targets.
And https://github.com/profianinc/steward/issues/66

File: deny.toml (hunk 9 above)
Comment (rjzak) [reply deny.toml]: Done

File: src/main.rs (hunk 10 above)
Comment (rjzak) [reply src/main.rs]: Done

File: src/main.rs (hunk 11 above)
Comment (rjzak) [reply src/main.rs]: Done

File: src/main.rs (hunk 12 above)
Comment (rjzak) [reply src/main.rs]: Done

File: src/main.rs (hunk 13 above)
Comment (rjzak) [reply src/main.rs]: Done

File: src/main.rs (hunk 14 above)
Comment (rjzak) [reply src/main.rs]: Moved, hopefully to the correct spot

**Review by rvolosatovs — APPROVED** [review]
